[Attributor] Take the address space from addrspacecast directly

shiltian opened the pull request:
Currently `AAAddressSpace` relies on identifying the address spaces of all underlying objects. However, it might infer sub-optimal address space when the underlying object is a function argument. In `AMDGPUPromoteKernelArgumentsPass`, the promotion of a pointer kernel argument is by adding a series of `addrspacecast` instructions (as shown below), and hoping `InferAddressSpacePass` can pick it up and do the rewriting accordingly.

Before promotion:

```
define amdgpu_kernel void @kernel(ptr %to_be_promoted) {
  %val = load i32, ptr %to_be_promoted
  ...
  ret void
}
```

After promotion:

```
define amdgpu_kernel void @kernel(ptr %to_be_promoted) {
  %ptr.cast.0 = addrspace cast ptr % to_be_promoted to ptr addrspace(1)
  %ptr.cast.1 = addrspace cast ptr addrspace(1) %ptr.cast.0 to ptr
  # all the use of %to_be_promoted will use %ptr.cast.1
  %val = load i32, ptr %ptr.cast.1
  ...
  ret void
}
```

When `AAAddressSpace` analyzes the code after promotion, it will take `%to_be_promoted` as the underlying object of `%ptr.cast.1`, and use its address space (which is 0) as its final address space, thus simply do nothing in `manifest`. The attributor framework will them eliminate the address space cast from 0 to 1 and back to 0, and replace `%ptr.cast.1` with `%to_be_promoted`, which basically reverts all changes by `AMDGPUPromoteKernelArgumentsPass`.

IMHO I'm not sure if `AMDGPUPromoteKernelArgumentsPass` promotes the argument in a proper way. To improve the handling of this case, this PR adds an extra handling when iterating over all underlying objects. If an underlying object is a function argument, it means it reaches a terminal such that we can't futher deduce its underlying object further. In this case, we check all uses of the argument. If they are all `addrspacecast` instructions and their destination address spaces are same, we take the destination address space.

Fixes: SWDEV-482640.

shiltian commented on llvm/test/CodeGen/AMDGPU/addrspacecast.ll:
```diff
@@ -215,11 +215,14 @@ define amdgpu_kernel void @use_flat_to_constant_addrspacecast(ptr %ptr) #0 {
 }
 
 ; HSA-LABEL: {{^}}cast_0_group_to_flat_addrspacecast:
+; CI: s_load_dword [[APERTURE:s[0-9]+]], s[4:5], 0x10
```
I just recover these lines from the commit before `AAAddressSpace` was enabled in `AMDGPUAttributorPass`.

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12571,12 +12571,37 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
 
   ChangeStatus updateImpl(Attributor &A) override {
-    int32_t OldAddressSpace = AssumedAddressSpace;
+    uint32_t OldAddressSpace = AssumedAddressSpace;
```
Need to do this inter-procedurally as well.

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12571,17 +12571,59 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
         return true;
+      // If an argument in generic address space has addrspace cast uses, and
+      // those casts are same, then we take the dst addrspace.
+      if (auto *Arg = dyn_cast<Argument>(&Obj)) {
```
I'm wondering if it is safe for us to directly say here that, if it is an argument of an entry point function, it is in global address space.

arsenm replied:
I don't think you can say that generically with the information available in a non-target pass. I would also consider this case a frontend bug, and they should emit parameters with the correct address space to begin with. We already do some hacking around clang using flat pointers in the backend

shiltian replied:
Well, I'd say that as well, but this is exactly what we do in the kernel argument promotion pass. I have conservative options on whether the pass should do that way.

shiltian replied:
I talked to @rampitec that we can't change kernel's signature. We can only recognize the pattern and support it.

arsenm replied:
Of course you can change the kernel signature. clang should emit IR with addrspace(1) values directly.

shiltian replied:
Per discussion with @rampitec, in HIP all pointers which can be passed to a kernel are flat.

arsenm replied:
At a source level, yes. But that has no bearing on what IR clang produces. clang can and should emit ptr addrspace(1) pointers with a cast to addrspace(0)

shiltian replied:
It looks like for HIP we already emit the kernel with AS 1 pointer arguments. However, is it always the case?

```
#include <cstdint>
#include <hip/hip_runtime.h>

__device__ __constant__ int constNumber[4] = {1, 2, 3, 4};

__global__ void kernel(int *out, int *in) { *out = in[3]; }

int main(int argc, char *argv[]) {
  int out;

  int *out_dev = nullptr;
  hipError_t err = hipMalloc(&out_dev, sizeof(int));
  if (err != hipSuccess)
    return 1;

  kernel<<<1, 1>>>(out_dev, constNumber);

  err = hipMemcpyDtoH(&out, out_dev, sizeof(int));
  if (err != hipSuccess)
    return 2;

  printf("out=%d\n", out);

  return 0;
}
```

The compiler doesn't complain anything, but at runtime it crashes due to memory access fault. The IR shows the kernel signature is:

```
define protected amdgpu_kernel void @_Z6kernelPiS_(ptr addrspace(1) noundef %out.coerce, ptr addrspace(1) noundef %in.coerce)
```

shiltian replied:
If that's the case, the `AMDGPUPromoteKernelArgumentsPass` should not exist at the first place.
Both Flang and Clang (for OpenMP target offloading) still generate generic pointer.
HIP and OpenCL front ends generate AS1 pointer, which is convenient.

arsenm replied:
Yes, it shouldn't. It's a hack for frontends not emitting the nicer IR

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12571,17 +12571,59 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
+      // If an argument in generic address space has addrspace cast uses, and
+      // those casts are same, then we take the dst addrspace.
+      if (auto *Arg = dyn_cast<Argument>(&Obj)) {
+        if (Arg->getType()->getPointerAddressSpace() == 0) {
```
should query the target generic address space and not assume 0

shiltian replied:
Well, I understand I should use `TargetTransformInfo::getFlatAddressSpace`, but the fact is it doesn't always work. It requires to pass a function to query it.  Here, the associated value can be a `Constant` which doesn't associate with any function, therefore I can't get TTI. It's quite unfortunate that the flat address space information is part of a function analysis result. Is there any way to work around it? I have an ugly solution, which is to pick a function from the module randomly, query the flat address space, and store it in the information cache. I don't really want to do that.

arsenm replied:
getFlatAddressSpace really belongs in TargetMachine, not TTI

shiltian replied:
I posted a [PR](https://github.com/llvm/llvm-project/pull/108594) to add it in `TargetMachine`, then I run into another issue: `TargetMachine` is also not available here. Unlike `AMDGPUAttributor`, AAs in `llvm/lib/Transforms/IPO/AttributorAttributes.cpp` as well as their user passes, may not have access to `TargetMachine`.

arsenm commented on llvm/test/CodeGen/AMDGPU/aa-as-infer.ll:
```diff
@@ -243,3 +243,38 @@ define void @foo(ptr addrspace(3) %val) {
+  ret void
+}
+
+define internal void @use_kernel_argument_after_promotion(ptr %p, i32 %val) {
```
Function names mention kernels but there are no kernels here

shiltian replied:
K, I'll remove the kernel because the callers just mention it's the kernel argument pattern, not really kernel argument.

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12571,8 +12571,35 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
+    // We assume the source address space is the correct one.
+    Value *V = &getAssociatedValue();
+    if (auto *ASCI = dyn_cast<AddrSpaceCastInst>(V)) {
+      assert(ASCI->getDestAddressSpace() == 0 &&
```
Don't hardcoded the generic = 0, query TTI

shiltian replied:
See comment below or above.

arsenm commented:
> If the value to be analyzed is directly from addrspacecast, we take the source
address space directly. 

I don't think this is valid in general. You are allow to speculatively produce invalid addrspacecasts. For example:

```
__generic int* g_ptr = ...;
__local int* l_ptr = (__local int*) g_ptr;
if (is_shared(g_ptr)) 
  *l_ptr = 1;


```

shiltian commented:
> > If the value to be analyzed is directly from addrspacecast, we take the source
> > address space directly.
> 
> I don't think this is valid in general. You are allow to speculatively produce invalid addrspacecasts. For example:
> 
> ```
> __generic int* g_ptr = ...;
> __local int* l_ptr = (__local int*) g_ptr;
> if (is_shared(g_ptr)) 
>   *l_ptr = 1;
> ```

That doesn't matter because the scope is the instruction. In your example here, the pointer of the store is already in the specific address space. That will not change no matter whether the branch is taken or not. `AAAddressSpace` doesn't say `g_ptr` is in what address space. It just changes the pointer operand of memory instruction.

shiltian commented:
I unstacked from https://github.com/llvm/llvm-project/pull/108786 to unblock this since the ticket needs to be fixed promptly.

shiltian commented on llvm/test/CodeGen/AMDGPU/aa-as-infer.ll:
```diff
@@ -243,3 +243,38 @@ define void @foo(ptr addrspace(3) %val) {
+  %p.cast.1 = addrspacecast ptr addrspace(1) %p.cast.0 to ptr
+  call void @use_argument_after_promotion(ptr %p.cast.1, i32 %val)
+  ret void
+}
```
what do you mean by "vector of pointers"?

shiltian replied:
I'm not sure if this is needed because it can't be handled by `amdgpu-promote-kernel-arguments`.

arsenm replied:
It should be handeable. It doesn't matter if it's handled or not, a test still should stress these cases to ensure it doesn't crash.

shiltian commented:
bump

shiltian committed:
[Attributor] Take the address space from addrspacecast directly

If the value to be analyzed is directly from addrspacecast, we take the source
address space directly. This is to improve the case where in
`AMDGPUPromoteKernelArgumentsPass`, the kernel argument is promoted by
insertting an addrspacecast directly from a generic pointer. However, during the
analysis, the underlying object will be the generic pointer, instead of the
addrspacecast, thus the inferred address space is the generic one, which is not
ideal.

shiltian commented on llvm/test/CodeGen/AMDGPU/aa-as-infer.ll:
```diff
@@ -243,3 +243,88 @@ define void @foo(ptr addrspace(3) %val) {
+  ret void
+}
+
+define void @vec_kernel_argument_promotion_pattern_intra_procedure(<2 x ptr> %p, i32 %val) {
```
@arsenm I added the test here but they don't make too much sense. The AA only works for load/store like instructions when a pointer is used as an operand. Apparently `<2 x ptr>` is not a valid operand for load/store like instructions.

arsenm replied:
Yes, this pass is much more limited compared to InferAddressSpaces by only operating per use (and only for specific uses). Plus this will produce a lot more spurious casts

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12583,16 +12583,37 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
   }
 
   ChangeStatus updateImpl(Attributor &A) override {
+    assert(A.getInfoCache().getFlatAddressSpace().has_value());
```
This assert is redundant, the attempt to use it will fail anyway (plus I still think it's nicer to not use optional, and just let the invalid address space value naturally not do anything)
